src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -171,35 +171,27 @@ internal static unsafe void WriteGatherAtOffset(SafeFileHandle handle, IReadOnly
                 stackalloc Interop.Sys.IOVector[IovStackThreshold] :
                 new Interop.Sys.IOVector[buffersCount];
 
-            try
+            long totalBytesToWrite = 0;
+            for (int i = 0; i < buffersCount; i++)
             {
-                int buffersOffset = 0, firstBufferOffset = 0;
-                while (true)
-                {
-                    long totalBytesToWrite = 0;
-
-                    for (int i = buffersOffset; i < buffersCount; i++)
-                    {
-                        ReadOnlyMemory<byte> buffer = buffers[i];
-                        totalBytesToWrite += buffer.Length;
-
-                        MemoryHandle memoryHandle = buffer.Pin();
-                        vectors[i] = new Interop.Sys.IOVector { Base = firstBufferOffset + (byte*)memoryHandle.Pointer, Count = (UIntPtr)(buffer.Length - firstBufferOffset) };
-                        handles[i] = memoryHandle;
+                ReadOnlyMemory<byte> buffer = buffers[i];
+                totalBytesToWrite += buffer.Length;
 
-                        firstBufferOffset = 0;
-                    }
-
-                    if (totalBytesToWrite == 0)
-                    {
-                        break;
-                    }
+                MemoryHandle memoryHandle = buffer.Pin();
```

Comment:
What happens if Pin throws an exception? Won't we now leak all the previously created handles?

Reply:
@stephentoub good catch, I forgot that the users can provide a custom `MemoryManager` and have `Pin` throw.